docs/root/intro/life_of_a_request.rst

```diff
@@ -0,0 +1,631 @@
+(``envoy.filters.listener.tls_inspector``). This filter examines the initial TLS handshake and
+extracts the server name (SNI). The SNI is then made available for filter chain matching. The TLS
+inspector does not appear explicitly in the listener filter chain configuration, instead Envoy
+inserts this automatically whenever there is a need for SNI (or ALPN) in a listener’s filter chain.
+
+.. image:: /_static/lor-listener-filters.svg
+   :width: 80%
+   :align: center
+
+The TLS inspector filter implements the :repo:`ListenerFilter <include/envoy/network/filter.h>`
+interface. All filter interfaces, whether listener or network/HTTP, require that filters implement
+callbacks for specific connection or stream events. In the case of `ListenerFilter`, this is:
+
+
+.. code-block:: cpp
+
+  virtual FilterStatus onAccept(ListenerFilterCallbacks& cb) PURE;
+
+``onAccept()`` allows a filter to run during the TCP accept processing. The ``FilterStatus``
+returned by the callback controls how the listener filter chain will continue. Listener filters may
+pause the filter chain and then later resume, e.g. in response to an RPC made to another service.
+
+Information extracted from the listener filters and connection properties is then used to match a
+filter chain, giving the network filter chain and transport socket that will be used to handle the
+connection.
+
+.. image:: /_static/lor-filter-chain-match.svg
+   :width: 50%
+   :align: center
+
+.. _life_of_a_request_tls_decryption:
+
+3. TLS transport socket decryption
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Envoy offers pluggable transport sockets via the
+:repo:`TransportSocket <include/envoy/network/transport_socket.h>`
+extension interface. Transport sockets follow the lifecycle events of a TCP connection and
+read/write into network buffers. Some key methods that transport sockets must implement are:
+
+.. code-block:: cpp
+
+  virtual void onConnected() PURE;
+  virtual IoResult doRead(Buffer::Instance& buffer) PURE;
+  virtual IoResult doWrite(Buffer::Instance& buffer, bool end_stream) PURE;
+  virtual void closeSocket(Network::ConnectionEvent event) PURE;
+
+When data is available on a TCP connection, ``Network::ConnectionImpl::onReadReady()`` invokes the
+:ref:`TLS <arch_overview_ssl>` transport socket via ``SslSocket::doRead()``. The transport socket
+then performs a TLS handshake on the TCP connection. When the handshake completes,
+``SslSocket::doRead()`` provides a decrypted plain text byte stream to an instance of
+``Network::FilterManagerImpl``, responsible for managing the network filter chain.
+
+.. image:: /_static/lor-transport-socket.svg
+   :width: 80%
+   :align: center
+
+It’s important to note that no operation, whether it’s a TLS handshake or a pause of a filter
+pipeline is truly blocking. Since Envoy is event-based, any situation in which processing requires
+additional data will lead to early event completion and yielding of the CPU to another event. When
```

Comment:
just wanted to say I really like this paragraph. Super helpful for people coming into event-based programming for the first time.